plugin/src/main/kotlin/org/jlleitschuh/gradle/ktlint/tasks/BaseKtLintCheckTask.kt

```diff
@@ -254,6 +255,12 @@ abstract class BaseKtLintCheckTask @Inject constructor(
             spec.classpath.from(ktLintClasspath, ruleSetsClasspath)
             spec.forkOptions { options ->
                 options.maxHeapSize = workerMaxHeapSize.get()
+
+                // Work around ktlint triggering reflective access errors from the embedded Kotlin
+                // compiler. See https://youtrack.jetbrains.com/issue/KT-43704 for details.
+                if (JavaVersion.current() >= JavaVersion.VERSION_16) {
```

Comment:
I am not that familiar with the  Worker API yet. are we sure this is the correct way to check the java versions?
ie. is it always inherited from the java version of the gradle process? or is there a toolchain/fork option we should be looking at?

Reply:
Good question. I don't think that, at this point, we are executing in the worker process. At this point we are still in the Gradle process, so it will be the JVM version of that process/classloader.

The workers are running in their own process/classloader, and static variables are unique-per-classloader hierarchy IIRC. As such, you can safely assume that any `JavaVersion.current()` calls will be accurate within the JVM classloader they are executing inside of.

Reply:
Looking at Gradle's exec code, the default `JavaForkOptions` is set up [here](https://github.com/gradle/gradle/blob/ab01e7fc05e4d14186b0a706e0737fe0b20f458f/subprojects/core/src/main/java/org/gradle/process/internal/DefaultExecActionFactory.java#L369-L375):

```java
@Override
public JavaForkOptionsInternal newDecoratedJavaForkOptions() {
    JavaDebugOptions javaDebugOptions = objectFactory.newInstance(DefaultJavaDebugOptions.class, objectFactory);
    final DefaultJavaForkOptions forkOptions = instantiator.newInstance(DefaultJavaForkOptions.class, fileResolver, fileCollectionFactory, javaDebugOptions);
    forkOptions.setExecutable(Jvm.current().getJavaExecutable());
    return forkOptions;
}
```

So by default, Gradle appears to set the java executable for the child process to the same executable as the Gradle daemon. To override this, I believe we would need to call `setExecutable` ourselves with a path to some other Java toolchain's `javac`. Since we don't do that, `JavaVersion.current()` should be accurate for both the Gradle daemon process and the forked worker process.